pkg/util/discardedseries/perlabelset_tracker.go

```diff
@@ -0,0 +1,140 @@
+package discardedseries
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	perLabelsetSeriesLimit = "per_labelset_series_limit"
+)
+
+type labelsetCounterStruct struct {
+	*sync.RWMutex
+	labelsetSeriesMap map[uint64]*seriesCounterStruct
+}
+
+type DiscardedSeriesPerLabelsetTracker struct {
+	*sync.RWMutex
+	userLabelsetMap                 map[string]*labelsetCounterStruct
+	discardedSeriesPerLabelsetGauge *prometheus.GaugeVec
+}
+
+func NewDiscardedSeriesPerLabelsetTracker(discardedSeriesPerLabelsetGauge *prometheus.GaugeVec) *DiscardedSeriesPerLabelsetTracker {
+	tracker := &DiscardedSeriesPerLabelsetTracker{
+		RWMutex:                         &sync.RWMutex{},
+		userLabelsetMap:                 make(map[string]*labelsetCounterStruct),
+		discardedSeriesPerLabelsetGauge: discardedSeriesPerLabelsetGauge,
+	}
+	return tracker
+}
+
+func (t *DiscardedSeriesPerLabelsetTracker) Track(user string, series uint64, matchedLabelsetHash uint64, matchedLabelsetId string) {
+	t.RLock()
+	labelsetCounter, ok := t.userLabelsetMap[user]
+	t.RUnlock()
+	if !ok {
+		t.Lock()
+		labelsetCounter, ok = t.userLabelsetMap[user]
+		if !ok {
+			labelsetCounter = &labelsetCounterStruct{
+				RWMutex:           &sync.RWMutex{},
+				labelsetSeriesMap: make(map[uint64]*seriesCounterStruct),
+			}
+			t.userLabelsetMap[user] = labelsetCounter
+		}
+		t.Unlock()
+	}
+
+	labelsetCounter.RLock()
+	seriesCounter, ok := labelsetCounter.labelsetSeriesMap[matchedLabelsetHash]
+	labelsetCounter.RUnlock()
+	if !ok {
+		labelsetCounter.Lock()
+		seriesCounter, ok = labelsetCounter.labelsetSeriesMap[matchedLabelsetHash]
+		if !ok {
+			seriesCounter = &seriesCounterStruct{
+				RWMutex:        &sync.RWMutex{},
+				seriesCountMap: make(map[uint64]struct{}),
+				labelsetId:     matchedLabelsetId,
+			}
+			labelsetCounter.labelsetSeriesMap[matchedLabelsetHash] = seriesCounter
+		}
+		labelsetCounter.Unlock()
+	}
+
+	seriesCounter.RLock()
+	_, ok = seriesCounter.seriesCountMap[series]
+	seriesCounter.RUnlock()
+	if !ok {
+		seriesCounter.Lock()
+		_, ok = seriesCounter.seriesCountMap[series]
+		if !ok {
+			seriesCounter.seriesCountMap[series] = struct{}{}
+		}
+		seriesCounter.Unlock()
+	}
+}
+
+func (t *DiscardedSeriesPerLabelsetTracker) UpdateMetrics() {
+	usersToDelete := make([]string, 0)
+	labelsetsToDelete := make([]uint64, 0)
+	t.RLock()
+	for user, labelsetCounter := range t.userLabelsetMap {
+		labelsetCounter.RLock()
+		if len(labelsetCounter.labelsetSeriesMap) == 0 {
+			usersToDelete = append(usersToDelete, user)
+		}
+		for labelsetHash, seriesCounter := range labelsetCounter.labelsetSeriesMap {
+			seriesCounter.Lock()
+			count := len(seriesCounter.seriesCountMap)
+			t.discardedSeriesPerLabelsetGauge.WithLabelValues(perLabelsetSeriesLimit, user, seriesCounter.labelsetId).Set(float64(count))
```

Comment:
Why are we hardcoding the reason. We didn't hardcode it to work for per labelset discarded reason only for discarded samples.